### PR TITLE
[AutoDiff] [stdlib] Add 'EuclideanDifferentiable' protocol.

### PR DESCRIPTION
**Pull request by rxwei**
Some iterative mathematical optimization algorithms (e.g. gradient descent with weight decay), when applied on a model (of type `Model` that conforms to `Differentiable`), rely on computing tangent vectors (of type `Model.TangentVector`) based on a gradient and an existing parameter (a stored property of `Model`). However, models that are differentiable via Euclidean differentiation are not necessarily vector spaces such that the `Model` equals `Model.TangentVector`, because `Model` may contain properties that do not appear in `Model.TangentVector`, such as Boolean flags or other hyperparameters. This use case is extremely common in machine learning, where most differentiation happens in the Euclidean space.

To make this use case possible, we introduce `EuclideanDifferentiable`, which generalizes types that are a product manifold of a differentiable vector space and some arbitrary manifold where the product manifold's tangent space is equal to the differentiable vector space component. In other words, `T: EuclideanDifferentiable` means that `T == V × M` where `V: Differentiable & AdditiveArithmetic` and `T.TangentVector == V == V.TangentVector`. The protocol has a `vectorView` property, which returns the `V` part (the differentiable vector space part) of the value.

The `EuclideanDifferentiable` allows us to **generically** express the kind of special iterative optimization algorithms that use a models parameters when computing a tangent vector.

```swift
let 𝛁L: Model.TangentVector = ...
model.move(along: -η * 𝛁L - η * λ * model.vectorView))
```

TODO: We need to implement derived conformances for the `vectorView` property.

Thanks to @sgugger, @marcrasi, @mattjj, @dougalm, and @caseychu for their input on solving this use case.

Partially resolves [tensorflow/swift-apis#456](https://github.com/tensorflow/swift-apis/issues/456).

### CODE REVIEW

**Comment by rxwei** [comment]
Note: This will not be in the first Swift evolution proposal for the differentiable programming feature. It will likely be proposed in [Phase 3](https://docs.google.com/document/d/1bPepWLfRQa6CtXqKA8CDQ87uZHixNav-TFjLSisuKag/edit#bookmark=id.jrkwtcmp81ct).

**Review by dan-zheng — APPROVED** [review]
LGTM!
This solution fills the abstraction hole and is much cleaner than [`AllDifferentiableVariables`](https://github.com/apple/swift/pull/26527).
I'm a fan of the organization of short initial description + longer mathematical description.

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

File: stdlib/public/core/AutoDiff.swift [hunk 1]
```diff
@@ -183,6 +183,41 @@ public extension Differentiable where TangentVector == Self {
   }
 }
 
+/// A type that consists of a differentiable vector space and some other
+/// non-differentiable component.
+///
+/// Mathematically, this represents a product manifold that consists of
+/// a differentiable vector space and some arbitrary manifold, where the tangent
```
Comment (dan-zheng) [inline stdlib/public/core/AutoDiff.swift]: It seems non-controversial that types are isomorphic to sets/spaces.

Wikipedia suggests manifolds are special topological spaces:
image

Though this description seems to imply that all types can be interpreted as "`some arbitrary manifold`", including `Bool` and other wildly "non-differentiable" types. Is the description accurate?

Comment (dan-zheng) [reply stdlib/public/core/AutoDiff.swift]: Maybe rephrasing "some arbitrary manifold" as "some other arbitrary component" would be less controversial, as "component" is vague.

Comment (rxwei) [reply stdlib/public/core/AutoDiff.swift]: We've had the assumption already for the base `Differentiable` protocol, which allows us to include values of arbitrary types as stored properties and to call the whole type a "differentiable manifold". This should definitely be addressed if brought up in a Swift evolution review.

Comment (rxwei) [reply stdlib/public/core/AutoDiff.swift]: > Maybe rephrasing "some arbitrary manifold" as "some other arbitrary component" would be less controversial, as "component" is vague.

IIUC, the product of a manifold and a non-manifold is no longer a manifold. This would contradict even the base protocol where everything `Differentiable` is a differentiable manifold (hence a manifold).

Comment (dan-zheng) [reply stdlib/public/core/AutoDiff.swift]: > We've had the assumption already for the base Differentiable protocol, which allows us to include values of arbitrary types as stored properties and to call the whole type a "differentiable manifold".

That's a great point, thanks 🙂

File: stdlib/public/core/AutoDiff.swift [hunk 2]
```diff
@@ -183,6 +183,41 @@ public extension Differentiable where TangentVector == Self {
...
+///     @noDerivative var useBias: Bool
+/// }
+/// ```
+///
+/// - Note: Conform a type to `EuclideanDifferentiable` if it is differentiable
+///   only with respect to its vector space component and when its
+///   `TangentVector` is equal to its vector space component.
+public protocol EuclideanDifferentiable: Differentiable {
+  /// The differentiable vector component of `self`.
+  var vectorView: TangentVector { get set }
```
Comment (saeta) [inline stdlib/public/core/AutoDiff.swift]: I thought we were planning on renaming TangentVector to simply Tangent. Are we sure that vectorView is the right name here? (I think so, but just wanted to flag this. Better yet, could we mark this as unstable for now?)

Comment (rxwei) [reply stdlib/public/core/AutoDiff.swift]: We were never planning on renaming `TangnetVector` to `Tangent`. There are multiple reasons `TangentVector` was named this way: 1. Swift type names describe instances, not spaces; 2. domain-specific type names should be Google-able: "tangent vector" is while "tangent" is not.

Comment (rxwei) [reply stdlib/public/core/AutoDiff.swift]: To your question about the naming of `vectorView`, I am sure this is the right name after analyzing various options, bu like any names, this design is to be brought up in a design review. Marking this property as unable does not feel particularly useful since the entire feature is experimental.
